spring-cloud-gcp-data-datastore/src/main/java/org/springframework/cloud/gcp/data/datastore/repository/query/PartTreeDatastoreQuery.java

```diff
@@ -125,50 +167,81 @@ else if (!returnedTypeIsNumber) {
+							getLimit(parameters));
+		}
+
+		Iterable rawResults = getDatastoreTemplate()
+				.queryKeysOrEntities(applyQueryBody(parameters, structredQueryBuilder, total), this.entityType);
+
+		Object result = StreamSupport.stream(rawResults.spliterator(), false).map(mapper).collect(collector);
 
 		if (this.tree.isDelete()) {
 			deleteFoundEntities(returnedTypeIsNumber, rawResults);
 		}
 
 		return this.tree.isExistsProjection() || isCountingQuery ? result
-				: convertResultCollection(result);
+				: convertResultCollection(result, collectionType);
+	}
+
+	private Integer getLimit(Object[] parameters) {
+		ParameterAccessor paramAccessor = new ParametersParameterAccessor(getQueryMethod().getParameters(), parameters);
+		Integer limit = null;
+		if (paramAccessor.getPageable().isPaged()) {
+			limit = paramAccessor.getPageable().getPageSize();
+		}
+		return limit;
 	}
 
-	private Object convertResultCollection(Object result) {
-		return this.datastoreTemplate.getDatastoreEntityConverter().getConversions()
-				.convertOnRead(result, this.queryMethod.getCollectionReturnType(),
-						this.queryMethod.getReturnedObjectType());
+	@VisibleForTesting
+	protected Object convertResultCollection(Object result, Class<?> collectionType) {
+		return getDatastoreTemplate().getDatastoreEntityConverter().getConversions()
+				.convertOnRead(result, collectionType, getQueryMethod().getReturnedObjectType());
 	}
 
 	private void deleteFoundEntities(boolean returnedTypeIsNumber, Iterable rawResults) {
 		if (returnedTypeIsNumber) {
-			this.datastoreTemplate.deleteAllById(rawResults, this.entityType);
+			getDatastoreTemplate().deleteAllById(rawResults, this.entityType);
 		}
 		else {
-			this.datastoreTemplate.deleteAll(rawResults);
+			getDatastoreTemplate().deleteAll(rawResults);
 		}
 	}
 
 	private StructuredQuery applyQueryBody(Object[] parameters,
-			StructuredQuery.Builder builder) {
+			StructuredQuery.Builder builder, boolean total) {
+		ParameterAccessor paramAccessor = new ParametersParameterAccessor(getQueryMethod().getParameters(), parameters);
 		if (this.tree.hasPredicate()) {
 			applySelectWithFilter(parameters, builder);
 		}
 
 		Integer limit = null;
+		Integer offset = null;
 		if (this.tree.isExistsProjection()) {
 			limit = 1;
 		}
 		else if (this.tree.isLimiting()) {
 			limit = this.tree.getMaxResults();
 		}
-		this.datastoreTemplate.applyQueryOptions(
-				builder, new DatastoreQueryOptions(limit, null, this.tree.getSort()), this.datastorePersistentEntity);
+
+		Sort sort = this.tree.getSort();
+		if (getQueryMethod().getParameters().hasPageableParameter()) {
+			sort = sort.and(paramAccessor.getPageable().getSort());
+		}
+		if (getQueryMethod().getParameters().hasSortParameter()) {
+			sort = sort.and(paramAccessor.getSort());
+		}
+
+		if (paramAccessor.getPageable().isPaged() && !total) {
+			limit = paramAccessor.getPageable().getPageSize() + (isSliceQuery() ? 1 : 0);
```

Comment:
the `? 1 : 0` is a cute little trick that should have a comment to explain it haha.

That `sliceQuery` for slice queries absolutely depends on this +1 extra result to set the limit boolean is another reason to get rid of that method and that inner class and have everything related to slices in this class here.